File: src/parsing.py
import re
from typing import List, Dict, NamedTuple

# Pre-compiled regex patterns for better performance
# Pattern to match transition tags with optional attributes
# Matches: <tag attr="value">content</tag>
# Captures: tag name, attributes string, content
_TRANSITION_PATTERN = re.compile(r'<(\w+)([^>]*)>(.*?)</\1>', re.DOTALL)

# Pattern to match key="value" or key='value' attributes
_ATTRIBUTE_PATTERN = re.compile(r'(\w+)=["\']([^"\']*)["\']')


class Transition(NamedTuple):
    """Represents a transition tag parsed from agent output.
    
    Attributes:
        tag: The tag name (goto, reset, function, call, fork, result)
        target: The target filename (empty for result tag)
        attributes: Dictionary of tag attributes (e.g., {"return": "X.md"})
        payload: The text content between tags (for result tag)
    """
    tag: str
    target: str
    attributes: Dict[str, str]
    payload: str


def parse_transitions(output: str) -> List[Transition]:
    r"""Extract transition tags from output.
    
    Recognizes:
        <goto>FILE.md</goto>
        <reset>FILE.md</reset>
        <function return="NEXT.md">EVAL.md</function>
        <call return="NEXT.md">CHILD.md</call>
        <fork next="NEXT.md" item="foo">WORKER.md</fork>
        <result>...</result>
    
    Args:
        output: The agent output text to parse
        
    Returns:
        List of Transition objects found in the output
        
    Raises:
        ValueError: If any tag target contains path separators (/ or \)
    """
    transitions = []
    
    for match in _TRANSITION_PATTERN.finditer(output):
        tag_name = match.group(1)
        attrs_str = match.group(2).strip()
        content = match.group(3)
        
        # Only process known transition tags
        if tag_name not in ("goto", "reset", "function", "call", "fork", "result"):
            continue
        
        # Parse attributes
        attributes = _parse_attributes(attrs_str)
        
        # Extract target and payload
        if tag_name == "result":
            # Result tag: payload is the content, no target
            target = ""
            payload = content
        else:
            # Other tags: target is the content (filename), no payload
            target = content.strip()
            payload = ""
            
            # Validate target is not empty
            if not target:
                raise ValueError(
                    f"Tag <{tag_name}> has empty target. "
                    "Non-result tags must specify a target filename."
                )
            
            # Validate path safety: target must not contain / or \
            if "/" in target or "\\" in target:
                raise ValueError(
                    f"Path '{target}' contains path separator. "
                    "Tag targets must be filenames only, not paths."
                )
        
        transitions.append(Transition(
            tag=tag_name,
            target=target,
            attributes=attributes,
            payload=payload
        ))
    
    return transitions


def _parse_attributes(attrs_str: str) -> Dict[str, str]:
    """Parse HTML-style attributes from a string.
    
    Args:
        attrs_str: String like 'return="X.md" item="foo"'
        
    Returns:
        Dictionary of attribute names to values
    """
    attributes = {}
    
    if not attrs_str:
        return attributes
    
    for match in _ATTRIBUTE_PATTERN.finditer(attrs_str):
        key = match.group(1)
        value = match.group(2)
        attributes[key] = value
    
    return attributes


def validate_single_transition(transitions: List[Transition]) -> None:
    """Validate that exactly one transition exists.
    
    Args:
        transitions: List of Transition objects
        
    Raises:
        ValueError: If the list does not contain exactly one transition
    """
    if len(transitions) != 1:
        raise ValueError(
            f"Expected exactly one transition, found {len(transitions)}"
        )

File: internal/parsing/parsing.go
package parsing

import (
	"fmt"
	"regexp"
	"strings"
)

// openTagRe matches opening tags for any recognized transition type.
// Group 1: tag name, Group 2: attributes string (may be empty).
//
// Go's RE2 does not support backreferences, so we cannot enforce matching
// close tags in a single regex. Instead we find the opening tag here, then
// locate the corresponding closing tag with a plain string search.
var openTagRe = regexp.MustCompile(`<(call-workflow|function-workflow|fork-workflow|reset-workflow|goto|reset|function|call|fork|result)([^>]*)>`)

// attrRe matches key="value" or key='value' attribute pairs.
// Two alternatives enforce matching quote characters so key="val' is rejected.
var attrRe = regexp.MustCompile(`(\w+)=(?:"([^"]*)"|'([^']*)')`)

// Transition represents a single transition tag parsed from agent output.
type Transition struct {
	Tag        string
	Target     string            // filename; empty for result tags
	Attributes map[string]string // e.g. {"return": "NEXT.md"}
	Payload    string            // content between tags; non-empty for result tags only
}

// ParseTransitions extracts all recognized transition tags from output.
//
// Tags may appear anywhere in the text. Unknown XML-like tags are silently
// ignored. Returns an error if a non-result tag has an empty target or a
// target containing a path separator (/ or \).
func ParseTransitions(output string) ([]Transition, error) {
	var transitions []Transition

	for _, match := range openTagRe.FindAllStringSubmatchIndex(output, -1) {
		tagName := output[match[2]:match[3]]
		attrsStr := strings.TrimSpace(output[match[4]:match[5]])
		afterOpen := match[1] // byte position immediately after '>'

		closeTag := "</" + tagName + ">"
		closeIdx := strings.Index(output[afterOpen:], closeTag)
		if closeIdx < 0 {
			continue // no matching close tag; skip
		}
		content := output[afterOpen : afterOpen+closeIdx]

		attrs := parseAttributes(attrsStr)

		var target, payload string
		if tagName == "result" {
			payload = content
		} else {
			target = strings.TrimSpace(content)
			if target == "" {
				return nil, fmt.Errorf(
					"tag <%s> has empty target. Non-result tags must specify a target filename.",
					tagName,
				)
			}
			if !IsWorkflowTag(tagName) && strings.ContainsAny(target, "/\\") {
				return nil, fmt.Errorf(
					"path %q contains path separator. Tag targets must be filenames only, not paths.",
					target,
				)
			}
		}

		transitions = append(transitions, Transition{
			Tag:        tagName,
			Target:     target,
			Attributes: attrs,
			Payload:    payload,
		})
	}

	return transitions, nil
}

// parseAttributes parses HTML-style key="value" or key='value' pairs.
// The regex uses two alternatives to enforce matching quotes:
// group 2 captures the value for double-quoted attributes,
// group 3 captures the value for single-quoted attributes.
func parseAttributes(attrsStr string) map[string]string {
	attrs := make(map[string]string)
	for _, m := range attrRe.FindAllStringSubmatch(attrsStr, -1) {
		// m[2] is non-empty for double-quoted values; m[3] for single-quoted.
		value := m[2]
		if value == "" && m[3] != "" {
			value = m[3]
		}
		attrs[m[1]] = value
	}
	return attrs
}

// ValidateSingleTransition returns an error if transitions does not contain
// exactly one element. The caller is responsible for deciding how to respond
// (e.g. send a reminder prompt to the agent).
func ValidateSingleTransition(transitions []Transition) error {
	if len(transitions) != 1 {
		return fmt.Errorf("expected exactly one transition, found %d", len(transitions))
	}
	return nil
}

// IsWorkflowTag reports whether tag is a cross-workflow transition whose target
// is a workflow specifier (directory path or zip path) rather than a local
// state name. Workflow-tag targets are passed through to specifier.Resolve
// without being validated as filenames.
func IsWorkflowTag(tag string) bool {
	return tag == "call-workflow" || tag == "function-workflow" || tag == "fork-workflow" || tag == "reset-workflow"
}

// ExtractStateName strips the recognized state file extension (.md, .sh, .bat, .ps1)
// from filename, case-insensitively. If no recognized extension is present the
// filename is returned unchanged. Case of the base name is preserved.
func ExtractStateName(filename string) string {
	lower := strings.ToLower(filename)
	for _, ext := range []string{".md", ".sh", ".bat", ".ps1"} {
		if strings.HasSuffix(lower, ext) {
			return filename[:len(filename)-len(ext)]
		}
	}
	return filename
}
